core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -422,108 +462,52 @@ private[spark] class BlockManager(
       Option(
         shuffleBlockResolver.getBlockData(blockId.asInstanceOf[ShuffleBlockId]).nioByteBuffer())
     } else {
-      doGetLocal(blockId, asBlockResult = false).asInstanceOf[Option[ByteBuffer]]
+      blockInfoManager.lockForReading(blockId) match {
+        case None =>
+          logDebug(s"Block $blockId was not found")
+          None
+        case Some(info) =>
+          Some(doGetLocalBytes(blockId, info))
+      }
     }
   }
 
-  private def doGetLocal(blockId: BlockId, asBlockResult: Boolean): Option[Any] = {
-    blockInfoManager.lockForReading(blockId) match {
-      case None =>
-        logDebug(s"Block $blockId was not found")
-        None
-      case Some(info) =>
-        val level = info.level
-        logDebug(s"Level for block $blockId is $level")
-
-        // Look for the block in memory
+  private def doGetLocalBytes(blockId: BlockId, info: BlockInfo): ByteBuffer = {
+    val level = info.level
+    logDebug(s"Level for block $blockId is $level")
+    if (level.deserialized) {
```

Comment:
`level-deserialized` specifies that this block's in-memory values will be stored as Java objects. For a block at this storage level, there are three possibilities:

1. it's stored as java objects in memory
2. it's stored as bytes on disk
3. an error has occurred and the block is missing

The logic inside of this `if` case tries to check case 2 ahead of case 1 so that we avoid unnecessary serialization if a serialized copy already exists on disk.

Were you confused about the fact that we can call `getLocalBytes()` on a block which is cached as java objects?